Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LSPLogging
+import LanguageServerProtocol
+
+protocol DiagnosticReportManagerDelegate: AnyObject {
+  /// Notify the delegate that compute a report with a given snapshot
+  func reportWithSnapshotReceived(_ snapshot: DocumentSnapshot, compilerArgs: [String]) async throws
+    -> RelatedFullDocumentDiagnosticReport
+
+  /// Notify the delegate that compute a report with a given snapshot for fallback
+  func fallbackReportWithSnapshotReceived(_ snapshot: DocumentSnapshot) async throws
+    -> RelatedFullDocumentDiagnosticReport
+}
+
+actor DiagnosticReportManager {
+  /// The cache that stores reports for snapshot ids
+  ///
+  /// Conceptually, this is a dictionary. To prevent excessive memory usage we
+  /// only keep `cacheSize` entries within the array. Older entries are at the
+  /// end of the list, newer entries at the front.
+  private var reportCache:
+    [(
+      snapshotID: DocumentSnapshot.ID,
```

Comment:
The build setting should also be part of the key. If the build settings change, the diagnostics need to be re-computed. With the current implementation, we would just return the diagnostics for the document with the old build settings.

Reply:
good catch 👍 , let me fix it.